### PR TITLE
Initial Governance contract integration

### LINKED ISSUE #14: dBFT stage 4: integrate voting
Header chain should be verifiable on its own, but we don't want to include the set of not-often-changed keys into every block, so:
 * allow for a key list to be included into `extra` as well, but only use it when `nextConsensus` differs from the current set
 * it should be included into the genesis block
 * once it's in the header, take them, verify they equal the `nextConsensus` hash and use to verify the next block
 * CNs should get the list from the voting contract once it's ready

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Here's a draft that I have right now, several questions need to be decided, they are marked as TODOs in the code. Close #14.

~~The good news is that the code is able to call Governance API and get the current set of validators. The bad news is that sometimes contract call doesn't work as expected, the following error occurs on attempt to unpack the result, I'm investigating it.~~
UPD: the described problem is solved, now Governance calls work as expected. Now I'm working on the rest of less critical TODOs:
- [x] Decide about final contract hash (ref. https://github.com/bane-labs/go-ethereum/pull/146#discussion_r1519684969).
- [x] ~~Automate contract build and genesis allocations pre-filling.~~ Will be done in #149.
- [x] Make contract code network-independent (ref. https://github.com/bane-labs/go-ethereum/pull/146#discussion_r1519655264 and https://github.com/bane-labs/go-ethereum/pull/146#discussion_r1519662507) and update the rest of privnet/testnet configurations afterwards.
- [x] Deal with genesis-based contract call processing.
- [x] Deal with next epoch validators calculations (ref. https://github.com/bane-labs/go-ethereum/pull/146#discussion_r1519970256).
- [x] Add validators cache.

### CODE REVIEW

**Comment by AnnaShaleva** [comment]
@chenquanyu, @txhsl welcome to review, but please keep in mind that it's a very early version of integration.

@nicolegys, this PR is not ready for testing yet, I'll notify you as soon as it's ready.

**Comment by AnnaShaleva** [comment]
@roman-khimov, first version is ready for review. The contract version should be updated (currently the very old one is used to test 7-nodes privnet), but the idea of the PR won't be changed if no issues found.

File: consensus/dbft/dbft.go
Comment (chenquanyu) [inline consensus/dbft/dbft.go:1653]: Do we need to sort validators in contract? It seems to be enough to sort them here.

File: core/systemcontracts/contracts.go
Comment (chenquanyu) [inline core/systemcontracts/contracts.go]: May be we only need the interface abi here.

File: consensus/dbft/dbft.go
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go:1653]: > Do we need to sort validators in contract?

We don't. But I'd like to ensure, that it's OK for us to sort validators on dBFT engine side.

File: core/systemcontracts/contracts.go
Comment (AnnaShaleva) [reply core/systemcontracts/contracts.go]: Yes, it will be updated once we have Governance finalized.

File: consensus/dbft/dbft.go
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go:1653]: It's OK for us, so let this sorting code to be present here. TODO is removed.

File: core/state_processor.go
Comment (chenquanyu) [inline core/state_processor.go:201]: Now we perform a call to OnPerist method of Governance contract via direct state modification (without fully-qualified transaction creation; for free) before in-block transactions processing. I think we can raise an issue first and we can implement free transactions in the future.

Comment (AnnaShaleva) [reply core/state_processor.go:201]: Ref. #152.

**Comment by AnnaShaleva** [comment]
@roman-khimov, ready for review & merge, @nicolegys is performing 7-nodes privnet testing. Should we wait for test results before the merge?

File: consensus/dbft/dbft.go
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: TODOs in the final commit series?

File: consensus/dbft/dbft.go
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: TODOs in the final commit series?

File: consensus/dbft/dbft.go
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: Mixed with other changes.

File: consensus/dbft/dbft.go
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: Can we get this number from the contract?

File: consensus/dbft/dbft.go
Comment (roman-khimov) [inline consensus/dbft/dbft.go:456]: Whitespace.

File: consensus/dbft/dbft.go
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: Number?

File: internal/ethapi/api.go
Comment (roman-khimov) [inline internal/ethapi/api.go:1143]: Undocumented?

File: core/systemcontracts/contracts.go
Comment (roman-khimov) [inline core/systemcontracts/contracts.go]: You can do this in `init` and panic.

**Comment by roman-khimov** [comment]
We need to test voting that changes the set of validators, btw.

**Comment by nicolegys** [comment]
CNs failed when trying to change epoch.

```
CRIT [03-27|08:16:07.046] failed to process state from proposal    err="failed to process block: could not apply OnPersist [0x1774aa018de852e473c1d5cf332fc22ebfe7af960b8d7eb64370e9fa20c4354d]: onPersist call failed: execution reverted" number=12 "seal hash"=c98122..e91b71 "parent hash"=0x40f5b34ede642c91ea99563ef4f0bbe7d384604fa4c6d2db4c39df4e2e7bc517 "intermediate merkle root"=047e7d..96635e coinbase=0x1212000000000000000000000000000000000003 "gas limit"=30,000,000 "gas used"=0 difficulty=1 "mix digest"=0x0000000000000000000000000000000000000000000000000000000000000000 nonce=3 time=1,711,527,227 "uncle hash"=0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347 txs=0
```

**Comment by txhsl** [comment]
> CNs failed when trying to change epoch.

Have you tried to register at least 7 different addresses as candidates before epoch change? The related contract method is https://github.com/bane-labs/go-ethereum/blob/governance/contracts/solidity/Governance.sol#L119.

**Comment by AnnaShaleva** [comment]
> CNs failed when trying to change epoch.

Node expects the contract to properly handle `onPersist` call, and if it's not the case, then the node have to stop operation. The case is that it's Primary node who is trying to construct PrepareRequest, and if the construction fails, then we're toast.

> Have you tried to register at least 7 different addresses as candidates before epoch change?

So I'd expect the contract to handle the case when not enough candidates are registered. In this case StandBy validators should be returned. Because we will have such cases in the real network.

File: consensus/dbft/dbft.go
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: These TODOs are resolved in the subsequent commits. Fixed.

Just FYI, we have a set of similar commits in NeoGo :D

File: consensus/dbft/dbft.go
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: Fixed.

File: consensus/dbft/dbft.go
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: Moved to a separate commit.

File: consensus/dbft/dbft.go
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: It was a kind of an intermediate contract version, it's fixed in the later commits. But now when we have a relatively stable contract, we don't need a code for intermediate version, so removed.

File: consensus/dbft/dbft.go
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go:456]: Fixed.

File: consensus/dbft/dbft.go
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: [Number](https://github.com/bane-labs/go-ethereum/issues/154).

File: internal/ethapi/api.go
Comment (AnnaShaleva) [reply internal/ethapi/api.go:1143]: I followed the existing `DoCall` style, it also doesn't have a comment. But probably we may write better code. Fixed.

File: core/systemcontracts/contracts.go
Comment (AnnaShaleva) [reply core/systemcontracts/contracts.go]: Fixed.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**
